### PR TITLE
Android NDK camera support

### LINKED ISSUE #11952: C++ cv::VideoCapture.open(0) always return false on Android
##### System information (version)

- OpenCV => 3.4.1
- Operating System macOS High Sierra 10.13.6 / Platform => Android 7.1.1 API 25
- Compiler => Android NDK

###### Context

I'm writing a cross platform OpenCV based C++ library. The consuming code is a React Native Application through a [react native native module](https://facebook.github.io/react-native/docs/native-modules-android.html).

To be perfectly clear, there is no access from Java Code to C++ OpenCV on Android. There are events with the result of the OpenCV C++ code sent to Javascript through the react native bridge.

My native library is compiled on Android as a SHARED library. It is dynamically linked to the libopencv_world.so that is produced by the compilation of OpenCV C++ for Android.

###### What it does

Basically, it opens the device's default camera and take snapshots.

###### The outcome

This code is then ran on iOS and Android.

This is working perfectly well on iOS.
It fails on Android.

Here is the failing part of C++ code on Adndroid:

##### Steps to reproduce

```.cpp
// cap is a cv::VideoCapture object    
if (cap.open(0))
        {
            cap.set(cv::CAP_PROP_FRAME_WIDTH, CAM_WIDTH);
            cap.set(cv::CAP_PROP_FRAME_HEIGHT, CAM_HEIGHT);
        }
        else
        {
            reject("false", "cap.open(0) returned false");
        }
```

### PR DESCRIPTION
**Pull request by komakai**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work

This PR fixes #11952

Add an Android native camera capture backend to the `videoio` module.
Based loosely on https://github.com/android/ndk-samples/tree/main/camera
Requires Android native API level 24 or above (in order to access NDK camera APIs)

```
android_pack_config=ndk-18-api-level-24.config.py
```

### CODE REVIEW

**Review by alalek — APPROVED** [review]
Well done! Thank you for contribution :+1:

**Comment by jumpy88** [comment]
So, should now with this update `cv::VideoCapture.open(0)` succeed and return an handle to accede a camera on Android? Does it only work for internal cameras or also for USB cameras? (Camera2 API I understand this patch is based on should support USB cameras https://source.android.com/devices/camera/external-usb-cameras)

**Comment by komakai** [comment]
@jumpy88 it will work for any device that is accessible to the Android NDK Camera API (https://developer.android.com/ndk/reference/group/camera)

> (Camera2 API I understand this patch is based on should support USB cameras https://source.android.com/devices/camera/external-usb-cameras)

This link is aimed at device manufacturers who are building the Android OS itself, not general Android app developers. (Also note: Camera2 is a Java API whereas the OpenCV `videoio` module is written in C++)

**Comment by jumpy88** [comment]
@komakai Ops, sorry, I thought Camera2 was an API for developers and it was accessible in Android NDK. My knowledge about Android is still very poor. Ok, I'll give it a try with your code and will post the result.
Thank you for your contribution and support.

**Comment by jumpy88** [comment]
Hi, I've just done some experiments with 2e2df44eac79acb5bb124356d5e274e235b87641 commit in a Godot Engine proof of concept. It works with built-in cameras but doesn't seem to recognize external USB cameras. The only weird thing is sometimes the visualized frame appears having blue and red channels swapped, but I don't have enough information to open an issue (I don't know if it's due to OpenCV itself or is my fault). You can find the sample code here https://github.com/jumpy88/godot_ocv-camera-reader
Great job @komakai

File: modules/videoio/src/cap_android_camera.cpp [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+                desiredHeight = value;
+                settingHeight = true;
+                if (settingWidth && settingHeight) {
+                    setWidthHeight();
+                    settingWidth = false;
+                    settingHeight = false;
+                }
+                return true;
+            case CV_CAP_PROP_CONVERT_RGB:
+                convertToRgb = (value != 0);
```
Comment (alalek) [inline modules/videoio/src/cap_android_camera.cpp:381]: > `cv::cvtColor(yuv, out, convertToRgb ? cv::COLOR_YUV2RGB_YV12 : cv::COLOR_YUV2BGR_YV12);`

BTW, meaning of this property is different (name misleads unfortunately, it is not BGR / RGB switch):
- CV_CAP_PROP_CONVERT_RGB = 1 (default): VideoCapture returns frames as "BGR" 8UC3 cv::Mat.
- CV_CAP_PROP_CONVERT_RGB = 0: VideoCapture returns frames in other formats, like NV12, Y8, etc (format is retrieved and/or sometimes specified through `CAP_PROP_FOURCC`)

Comment (komakai) [reply modules/videoio/src/cap_android_camera.cpp:381]: @alalek OK thanks for the explanation - will prepare a fix for this when I get a minute.

Comment (splinterwhatever) [reply modules/videoio/src/cap_android_camera.cpp:381]: Is there a space in the api to allow for the BGR/RGB switch in addition to the correct convertToRGB behaviour?

cheers, Jeff

**Comment by Ay-De** [comment]
Hello,

I am using opencv 4.5.4. python reports on android: after calling 

cam = cv2.VideoCapture(0)
cam.isOpened(): True.

but i can't get a frame from the video stream. If i call 

retval, frame = cam.read()

retval is returned as False and Frame is None. Do i miss something? The code works on windows.

**Comment by Ay-De** [comment]
@komakai Do you know what i am missing here?

**Comment by Ay-De** [comment]
> @Ay-De running OpenCv Python on Android is not something I have tested. I would suggest to look in the Logcat output to see if there are any error messages there

Hello, i am using kivy to run python code on android. OpenCV is being imported, videocapture(0) (or 1) works also. As i said, the isOpened() function also returns true, but if i try to read a frame, with read(), that returns nothing. No error is thrown according to logcat, but the variable for frame says it is None.

**Comment by komakai** [comment]
@Ay-De Do you have an APK file you can share?

**Comment by Ay-De** [comment]
> @Ay-De Do you have an APK file you can share?

Yes, here:
https://github.com/Ay-De/Kivy_Python_Camera_Demo/releases/tag/0.0.17

There is also the source code (See the main.py in develop branch. sorry it is still quite messy). The APK is compiled with debug mode enabled.

**Comment by komakai** [comment]
OK - so I tried your APK and saw lots of output like this:

2022-02-14 21:30:53.148 26028-26132/org.ayde.kivycamera V/NativeCamera: session 0x7bce246ba0 capture completed
2022-02-14 21:30:53.149 26028-26066/org.ayde.kivycamera **E/NativeCamera: Unsupported format**

**Comment by Ay-De** [comment]
Ok sorry i somehow missed that part. I think the problem lies in returned image format from the sensor? Is there a way to specify it in OpenCV? OpenCV has a flag CAP_PROP_FORMAT to fetch undecoded RAW video streams, but unfortunatly, the android camera seems to support different ones and i can't currently find a way to set it on the Android side. Would it maybe be possible to set the camera format after opening the image stream by making direct API calls?
https://developer.android.com/reference/android/graphics/ImageFormat

I am sorry if i am derailing this thread here!

**Comment by komakai** [comment]
@Ay-De The error message is coming from this line:
https://github.com/opencv/opencv/blob/4.x/modules/videoio/src/cap_android_camera.cpp#L324

There were some bug fixes recently in that area of the code:
https://github.com/opencv/opencv/commit/68667d6057f6cd3fed80306c8845e03771a3f70a
https://github.com/opencv/opencv/commit/fa222b0ea27ebf03dd41132ee3da8c99ab4688fd
https://github.com/opencv/opencv/commit/21d0f4075141568f7bf6616c5f8e79c06987283e
So my guess would be the build you have is missing one or more of these fixes.

**Comment by Ay-De** [comment]
So, first of all, thanks for your help. I really appreciate it.

I updated my files to include opencv in version 4.5.5, which include the above mentioned fixes, but it still doesn't work. May i ask you, which tools you are using to get the above mentioned logs?

> 2022-02-14 21:30:53.148 26028-26132/org.ayde.kivycamera V/NativeCamera: session 0x7bce246ba0 capture completed
2022-02-14 21:30:53.149 26028-26066/org.ayde.kivycamera E/NativeCamera: Unsupported format

I am using logcat for debugging but i am not seeing this errors, i guess i am maybe filtering out to much? I would like to understand the problems myself better instead of just asking and for that i think i need a better insight.

I am currently only seeing these logs:

> 02-15 16:48:55.923 10556 10748 I python  : Run user program, change dir and execute entrypoint
02-15 16:48:55.966 10556 10748 I python  : [WARNING] [Config      ] Older configuration version detected (0 instead of 21)
02-15 16:48:55.966 10556 10748 I python  : [WARNING] [Config      ] Upgrading configuration in progress.
02-15 16:48:55.970 10556 10748 I python  : [INFO   ] [Logger      ] Record log in /data/user/0/org.ayde.kivycamera/files/app/.kivy/logs/kivy_22-02-15_0.txt
02-15 16:48:55.970 10556 10748 I python  : [INFO   ] [Kivy        ] v2.0.0
02-15 16:48:55.970 10556 10748 I python  : [INFO   ] [Kivy        ] Installed at "/data/user/0/org.ayde.kivycamera/files/app/_python_bundle/site-packages/kivy/__init__.pyc"
02-15 16:48:55.970 10556 10748 I python  : [INFO   ] [Python      ] v3.8.9 (default, Feb 15 2022, 16:29:31)
02-15 16:48:55.970 10556 10748 I python  : [Clang 9.0.9 (https://android.googlesource.com/toolchain/llvm-project a2a1e703c
02-15 16:48:55.970 10556 10748 I python  : [INFO   ] [Python      ] Interpreter at ""
02-15 16:48:56.256 10556 10748 I python  : [INFO   ] [Factory     ] 186 symbols loaded
02-15 16:48:56.341 10556 10748 I python  : [INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2 (img_pil, img_ffpyplayer ignored)
02-15 16:48:56.517 10556 10748 I python  : OpenCV version: 4.5.5
02-15 16:48:56.597 10556 10748 I python  : [INFO   ] [Window      ] Provider: sdl2
02-15 16:48:56.611 10556 10748 I python  : [INFO   ] [GL          ] Using the "OpenGL ES 2" graphics system
02-15 16:48:56.611 10556 10748 I python  : [INFO   ] [GL          ] Backend used <sdl2>
02-15 16:48:56.612 10556 10748 I python  : [INFO   ] [GL          ] OpenGL version <b'OpenGL ES 3.2 V@0530.0 (GIT@5a9022f91f, Ib11adbd47c, 1627309424) (Date:07/26/21)'>
02-15 16:48:56.612 10556 10748 I python  : [INFO   ] [GL          ] OpenGL vendor <b'Qualcomm'>
02-15 16:48:56.612 10556 10748 I python  : [INFO   ] [GL          ] OpenGL renderer <b'Adreno (TM) 660'>
02-15 16:48:56.613 10556 10748 I python  : [INFO   ] [GL          ] OpenGL parsed version: 3, 2
02-15 16:48:56.613 10556 10748 I python  : [INFO   ] [GL          ] Texture max size <16384>
02-15 16:48:56.613 10556 10748 I python  : [INFO   ] [GL          ] Texture max units <16>
02-15 16:48:56.631 10556 10748 I python  : [INFO   ] [Window      ] auto add sdl2 input provider
02-15 16:48:56.632 10556 10748 I python  : [INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
02-15 16:48:56.633 10556 10748 I python  : [INFO   ] [GL          ] NPOT texture support is available
02-15 16:48:57.073 10556 10748 I python  : Camera connected Status: True
02-15 16:48:57.478 10556 10748 I python  : Frame returned: False
02-15 16:48:57.478 10556 10748 I python  : Frame data: None
02-15 16:48:57.704 10556 10748 I python  : [INFO   ] [Text        ] Provider: sdl2
02-15 16:48:57.746 10556 10748 I python  : [WARNING] [Base        ] Unknown <android> provider
02-15 16:48:57.746 10556 10748 I python  : [INFO   ] [Base        ] Start application main loop

**Comment by komakai** [comment]
> May i ask you, which tools you are using to get the above mentioned logs?
In Android Studio look for a tab along the bottom called "Logcat"

There is an Android/Java sample here:
https://github.com/komakai/NdkCamtest/blob/master/app/src/main/java/net/telepathix/ndkcam/MainActivity.java#L97
If you copy the same logic into Python it should work 🤞

**Comment by Ay-De** [comment]
I finally got it to work. Thanks @komakai.

**Comment by asadasivam** [comment]
> Hi, I've just done some experiments with [2e2df44](https://github.com/opencv/opencv/commit/2e2df44eac79acb5bb124356d5e274e235b87641) commit in a Godot Engine proof of concept. It works with built-in cameras but doesn't seem to recognize external USB cameras. The only weird thing is sometimes the visualized frame appears having blue and red channels swapped, but I don't have enough information to open an issue (I don't know if it's due to OpenCV itself or is my fault). You can find the sample code here https://github.com/jumpy88/godot_ocv-camera-reader Thank you for the great job @komakai

Let us know any further improvements committed for external USB camera access? For external camera to support we need to enable below feature. My understanding is for some android mobile manufactures, USB Webcam access support is available, in that mobile CAMERA2 API will work. ie,. external camera will be detected in CameraManager API.

<uses-feature android:name="android.hardware.camera2.full" />

**Pull request merged**